Add Deprecations page

eksperimental opened the pull request:
First draft.
I need your input where you see the __"???"__

let me know what you all think

eksperimental commented:
You can see the live page formatted page here.
https://github.com/eksperimental/elixir/blob/deprecations/lib/elixir/pages/Deprecations.md

josevalim commented on lib/elixir/pages/Deprecations.md:
```diff
@@ -0,0 +1,57 @@
+# Deprecations
+
+## Deprecation policy
```
Shall we call it `## Policy`

josevalim commented on lib/elixir/pages/Deprecations.md:
```diff
@@ -0,0 +1,57 @@
+  3. The feature is removed. This can only happen on major releases. This means deprecated features in Elixir v1.x shall only be removed by Elixir v2.x.
+
+
+## Table of deprecated features: modules, functions, behaviours, protocols, typespecs, attributes and arguments
```
What about: `## Table of deprecated features` or `## Table of deprecations`?

josevalim commented on lib/elixir/pages/Deprecations.md:
```diff
@@ -0,0 +1,57 @@
+
+Deprecated feature        | Deprecated in | Replaced by             | Replacement available since
+:------------------------ | :------------ | :---------------------- | :---------------------------
+`GenEvent` module         | [v1.5]    | `Supervisor` and `GenServer`; `GenStage`; `:gen_event` | v1.0 (`Supervisor and GenServer`); v1.3 (GenStage); OTP v17.0 (`:gen_server`)
```
I believe you meant "OTP v17.0 (`:gen_event`)"

eksperimental replied:
oops

josevalim commented on lib/elixir/pages/Deprecations.md:
```diff
@@ -0,0 +1,57 @@
+:------------------------ | :------------ | :---------------------- | :---------------------------
+`GenEvent` module         | [v1.5]    | `Supervisor` and `GenServer`; `GenStage`; `:gen_event` | v1.0 (`Supervisor and GenServer`); v1.3 (GenStage); OTP v17.0 (`:gen_server`)
+`()` to mean `nil`        | [v1.5]    | `nil`                       | v1.0
+`@compile {:parse_transform, _}` in `Module` | [v1.5] | __???__     | __???__
```
??? = None

josevalim commented on lib/elixir/pages/Deprecations.md:
```diff
@@ -0,0 +1,57 @@
+`Task.find/2`             | [v1.3]        | Direct message matching | v1.0
+Non-map as second arg. in `URI.decode_query/2` | [v1.3] | Use a map | v1.0
+`Access` protocol         | [v1.1]        | `Access` behaviour      | v1.1
+`?\xHEX`                  | [v1.1]        | `0xHEX`                 | __???__ (EDITOR'S NOTE: this has been completely deprecated and it shouldn't have been until v2.0)
```
I believe we had announced it before 1.0 but we forgot to remove it on 1.0 (doing so only on 1.1).

eksperimental replied:
ok. thank you

eksperimental replied:
so the 0xHEX is available since v1.0, correct?

josevalim replied:
Yes!

josevalim commented on lib/elixir/pages/Deprecations.md:
```diff
@@ -0,0 +1,57 @@
+`Access.key/1`            | [v1.4]    | `Access.key/2`              | v1.3
+`Behaviour` module        | [v1.4]    | `@callback`                 | v1.0
+`Enum.uniq/2`             | [v1.4]    | `Enum.uniq_by/2`            | v1.2
+`Float.to_char_list/2`    | [v1.4]    | `:erlang.float_to_list/2`   | OTP v17.0
```
We should refer to OTP as "OTP 17". WIth no `v` and no minor version because we don't want developers to depend on specific features based on minor OTP versions.

josevalim commented on lib/elixir/pages/Deprecations.md:
```diff
@@ -0,0 +1,57 @@
+`IEx.Helpers.import_file/2` | [v1.4]  | `IEx.Helpers.import_file_if_available/2` | v1.3
+`Mix.Utils.underscore/1`  | [v1.4]    | `Macro.underscore/1`        | v1.2
+`Mix.Utils.camelize/1`    | [v1.4]    | `Macro.camelize/1`          | v1.2
+`Dict` behaviour          | [v1.3]    | __???__ `Keyword`; `Map`    | __???__ v1.0 (Keyword); v1.2 (Map)
```
This is correct (??? can be removed)

eksperimental replied:
[CHANGELOG for v1.2](https://github.com/elixir-lang/elixir/blob/v1.2/CHANGELOG.md#3-soft-deprecations-no-warnings-emitted) states that  states that Dict was soft-deprecated, but a warning is emitted, 
https://github.com/elixir-lang/elixir/commit/936b243517919c05b0391714dd22d8f56bde3f31#diff-98fc93be50711b2d5c492ef04d076f52R48

should we update CHANGELOG?

josevalim replied:
@eksperimental we warned only for those using Dict to implement their own dicts. The API itself was OK until later. So there are two deprecations: the Dict behaviour and the Dict API. The Dict behaviour has no replacement, as there is no longer a Dict thingy to implement, the Dict API is the one you mentioned above. So I would recommend changing "Dict behaviour" by "Dict module" on this section.

eksperimental replied:
exactly. so what I did was add

    `Dict` behaviour          | [v1.2]        | *None*                  | *None*

still feels weird not to suggest an alternative

whatyouhide commented on lib/elixir/pages/Deprecations.md:
```diff
@@ -0,0 +1,57 @@
+`Float.to_char_list/1`    | [v1.5]    | `Float.to_charlist/1`       | v1.3
+`Integer.to_char_list/1`  | [v1.5]    | `Integer.to_charlist/1`     | v1.3
+`Kernel.to_char_list/1`   | [v1.5]    | `Kernel.to_charlist/1`      | v1.3
+`String.to_char_list/1`   | [v1.5]    | `String.to_charlist/1`      | v1.3
```
There's the `:charlists` option in `Inspect.Opts` as well I believe.

whatyouhide commented on lib/elixir/pages/Deprecations.md:
```diff
@@ -0,0 +1,57 @@
+`Keyword.size/1`          | [v1.3]        | `Kernel.length/1`       | v1.0
+`Map.size/1`              | [v1.3]        | `Kernel.map_size/1`     | v1.0
+`/r` option in `Regex`    | [v1.3]        | `/U`                    | v1.1
+`Set` behaviour           | [v1.3]        | __???__                 | __???__
```
There's no replacement for the behaviour, now you should just use `MapSet`.

josevalim commented on lib/elixir/pages/Deprecations.md:
```diff
@@ -0,0 +1,57 @@
+`Mix.Utils.camelize/1`    | [v1.4]    | `Macro.camelize/1`          | v1.2
+`Dict` behaviour          | [v1.3]    | __???__ `Keyword`; `Map`    | __???__ v1.0 (Keyword); v1.2 (Map)
+`\x{X*}` inside strings/sigils/charlists | [v1.3] | `\uXXXX`; `\u{X*}` | v1.1
+Map/dictionary as second arg. in `Enum.group_by/3` | [v1.3] | *n/a* | *n/a*
```
in `Enum.group_by/2`. I believe we didn't have `group_by/3` by then.

josevalim replied:
What about using `None` instead of `n/a` which may not be clear in intent?

eksperimental replied:
@josevalim We have had `Enum.group_by/3` since v1.0

josevalim commented on lib/elixir/pages/Deprecations.md:
```diff
@@ -0,0 +1,57 @@
+Non-map as second arg. in `URI.decode_query/2` | [v1.3] | Use a map | v1.0
+`Access` protocol         | [v1.1]        | `Access` behaviour      | v1.1
+`?\xHEX`                  | [v1.1]        | `0xHEX`                 | __???__ (EDITOR'S NOTE: this has been completely deprecated and it shouldn't have been until v2.0)
+`as: true \| false` in `alias/2` and `require/2` | [v1.1]           | __???__ *n/a* | *n/a*
```
??? = None

josevalim commented:
This looks great, I have added some feedback. :)

eksperimental commented on lib/elixir/pages/Deprecations.md:
```diff
@@ -11,44 +11,49 @@ Elixir deprecations happens in 3 steps:
 `String.valid_character?/1` | [v1.3]      | `String.valid?/1`       | v1.0
 `Task.find/2`             | [v1.3]        | Direct message matching | v1.0
 Non-map as second arg. in `URI.decode_query/2` | [v1.3] | Use a map | v1.0
+`Dict` behaviour          | [v1.2]        | *None*                  | *None*
```
This has been added, please review.

eksperimental commented on lib/elixir/pages/Deprecations.md:
```diff
@@ -11,44 +11,49 @@ Elixir deprecations happens in 3 steps:
 `Mix.Utils.underscore/1`  | [v1.4]    | `Macro.underscore/1`        | v1.2
 `Mix.Utils.camelize/1`    | [v1.4]    | `Macro.camelize/1`          | v1.2
-`Dict` behaviour          | [v1.3]    | __???__ `Keyword`; `Map`    | __???__ v1.0 (Keyword); v1.2 (Map)
+`Dict` module             | [v1.3]    | `Keyword`; `Map`            | v1.0 (`Keyword`); v1.2 (`Map`)
```
This has been changed from "behaviour" to "module". please review.

eksperimental commented on lib/elixir/pages/Deprecations.md:
```diff
@@ -11,44 +11,49 @@ Elixir deprecations happens in 3 steps:
-`Float.to_string/2`       | [v1.4]    | `:erlang.float_to_binary/2` | OTP v17.0
+`Float.to_char_list/2`    | [v1.4]    | `:erlang.float_to_list/2`   | OTP 17
+`Float.to_string/2`       | [v1.4]    | `:erlang.float_to_binary/2` | OTP 17
+`HashDict` module         | [v1.4]    | `Map`                       | v1.2
```
This was not documented. And it has been added, please review.

eksperimental commented on lib/elixir/pages/Deprecations.md:
```diff
@@ -11,44 +11,49 @@ Elixir deprecations happens in 3 steps:
+`Float.to_char_list/2`    | [v1.4]    | `:erlang.float_to_list/2`   | OTP 17
+`Float.to_string/2`       | [v1.4]    | `:erlang.float_to_binary/2` | OTP 17
+`HashDict` module         | [v1.4]    | `Map`                       | v1.2
+`HashSet` module          | [v1.4]    | `MapSet`                    | v1.1
```
This was not documented. And it has been added, please review.

eksperimental commented on lib/elixir/pages/Deprecations.md:
```diff
@@ -11,44 +11,49 @@ Elixir deprecations happens in 3 steps:
 `Kernel.to_char_list/1`   | [v1.5]    | `Kernel.to_charlist/1`      | v1.3
 `String.to_char_list/1`   | [v1.5]    | `String.to_charlist/1`      | v1.3
-`t:char_list/0` typespec  | [v1.5]    | `t:charlist/0` typespec     | v1.3
+`:char_lists` key in `t:Inspect.Opts.t/0` type | [v1.5] | `:charlists` | v1.3
```
This was not documented. And it has been added, please review.

eksperimental commented on lib/elixir/pages/Deprecations.md:
```diff
@@ -11,44 +11,49 @@ Elixir deprecations happens in 3 steps:
 `String.to_char_list/1`   | [v1.5]    | `String.to_charlist/1`      | v1.3
-`t:char_list/0` typespec  | [v1.5]    | `t:charlist/0` typespec     | v1.3
+`:char_lists` key in `t:Inspect.Opts.t/0` type | [v1.5] | `:charlists` | v1.3
+`:as_char_lists` value in `t:Inspect.Opts.t/0` type | [v1.5] | `:as_charlists` | v1.3
```
This was not documented. And it has been added, please review.

eksperimental commented on lib/elixir/pages/Deprecations.md:
```diff
@@ -0,0 +1,62 @@
+:------------------------ | :------------ | :---------------------- | :---------------------------
+`Atom.to_char_list/1`     | [v1.5]    | `Atom.to_charlist/1`        | v1.3
+`Float.to_char_list/1`    | [v1.5]    | `Float.to_charlist/1`       | v1.3
+`GenEvent` module         | [v1.5]    | `Supervisor` and `GenServer`; `GenStage`; `:gen_event` | v1.0 (`Supervisor` and `GenServer`); v1.3 (`GenStage`); OTP 17 (`:gen_event`)
```
Should we link to GenStage and Erlang docs for :gen_event ?

whatyouhide replied:
I would say yes to these two because they are external to Elixir and it could help discoverability to link them.

eksperimental replied:
so for GenStage, should we link to Github repo or hex package, or hexdocs page?

whatyouhide replied:
I think the right place is hexdocs.pm.

eksperimental replied:
Yes. It looks like the Hex package page is the place to link to.

eksperimental commented on lib/elixir/pages/Deprecations.md:
```diff
@@ -0,0 +1,62 @@
+`HashSet` module          | [v1.4]    | `MapSet`                    | v1.1
+`Set` module              | [v1.4]    | `MapSet`                    | v1.1
+`Stream.uniq/2`           | [v1.4]    | `Stream.uniq_by/2`          | v1.2
+`IEx.Helpers.import_file/2` | [v1.4]  | `IEx.Helpers.import_file_if_available/1` | v1.3
```
Should we link to the docs page in hexdocs.pm?

whatyouhide replied:
I don't think we should link all of these, they're quite a lot. :| Did you mean linking all of them?

eksperimental replied:
nope. I mean those for other Elixir apps, such as IEx, which do not get linked automatically.

whatyouhide commented on lib/elixir/pages/Deprecations.md:
```diff
@@ -0,0 +1,62 @@
+`Kernel.to_char_list/1`   | [v1.5]        | `Kernel.to_charlist/1`  | v1.3
+`String.to_char_list/1`   | [v1.5]        | `String.to_charlist/1`  | v1.3
+`()` to mean `nil`        | [v1.5]        | `nil`                   | v1.0
+`:as_char_lists` value in `t:Inspect.Opts.t/0` type | [v1.5] | `:as_charlists` | v1.3
```
`:as_char_lists` value -> `:as_char_lists` field

eksperimental replied:
here, :as_char_lists is a value, not a key/field

josevalim replied:
:+1:

whatyouhide replied:
Ops, missed that. Sorry!

whatyouhide commented on lib/elixir/pages/Deprecations.md:
```diff
@@ -0,0 +1,62 @@
+`String.to_char_list/1`   | [v1.5]        | `String.to_charlist/1`  | v1.3
+`()` to mean `nil`        | [v1.5]        | `nil`                   | v1.0
+`:as_char_lists` value in `t:Inspect.Opts.t/0` type | [v1.5] | `:as_charlists` | v1.3
+`:char_lists` key in `t:Inspect.Opts.t/0` type | [v1.5] | `:charlists` | v1.3
```
key ->field

whatyouhide commented on lib/elixir/pages/Deprecations.md:
```diff
@@ -0,0 +1,62 @@
+`()` to mean `nil`        | [v1.5]        | `nil`                   | v1.0
+`:as_char_lists` value in `t:Inspect.Opts.t/0` type | [v1.5] | `:as_charlists` | v1.3
+`:char_lists` key in `t:Inspect.Opts.t/0` type | [v1.5] | `:charlists` | v1.3
+`char_list/0` type        | [v1.5]        | `charlist/0` type       | v1.3
```
Should this be linked with `t:`?

eksperimental replied:
since it's built in, it doesn't link to anything and it leaves the `t:`.
I think it's an ExDoc bug. While it doesn't link, it should still remove the `t:`

josevalim replied:
:+1:

eksperimental commented:
@josevalim @whatyouhide could you please review the lines commented in https://github.com/elixir-lang/elixir/pull/5808#pullrequestreview-23897280
the code is outdated, but the lines are still valid. it is just rows that has been introduced after the first commit

josevalim commented on lib/elixir/pages/Deprecations.md:
```diff
@@ -0,0 +1,62 @@
+`Dict` module             | [v1.3]        | `Keyword`; `Map`        | v1.0 (`Keyword`); v1.2 (`Map`)
+`Keyword.size/1`          | [v1.3]        | `Kernel.length/1`       | v1.0
+`Map.size/1`              | [v1.3]        | `Kernel.map_size/1`     | v1.0
+`Set` behaviour           | [v1.3]        | *None*                  | *None*
```
Theoretically the replacement for `Set` behaviour is to use `MapSet` directly.

josevalim commented on lib/elixir/pages/Deprecations.md:
```diff
@@ -0,0 +1,62 @@
+`Set` behaviour           | [v1.3]        | *None*                  | *None*
+`String.valid_character?/1` | [v1.3]      | `String.valid?/1`       | v1.0
+`Task.find/2`             | [v1.3]        | Direct message matching | v1.0
+`:append_first` option in `Kernel.defdelegate/2` | [v1.3] | *None*  | *None*
```
Theoretically the replacement is defining the function directly.

josevalim commented on lib/elixir/pages/Deprecations.md:
```diff
@@ -0,0 +1,62 @@
+`Access` protocol         | [v1.1]        | `Access` behaviour      | v1.1
+`as: true \| false` in `alias/2` and `require/2` | [v1.1] | *None*  | *None*
+`?\xHEX`                  | [v1.1]        | `0xHEX`                 | v1.0
+Empty string in `String.starts_with?/2`, `String.ends_with?/2`, `String.contains?/2`. *__NOTE__: Feature made back available in v1.3* | [v1.1] to [v1.2] | *None* | *None*
```
Theoretically the replacement is to explicitly check for `""` for before hand.

josevalim commented:
@eksperimental do we plan to include syntax deprecations? Such as `local_fun_as_variable` and `fn -> end`?

eksperimental commented:
yes, I think we should. Anything that emits a warning should be listed here,

what is the `fn -> end` deprecation?

josevalim commented:
```iex
iex(1)> fn -> end
warning: an expression is always required on the right side of ->. Please provide a value after ->
  iex:1
```

eksperimental commented:
I have addressed all the suggestions,
Please have a look at the "None"s and let me know if there is an alternative.

Also I am thinking that there are more syntax deprecations to add. What we could do is to split the table in two. One for Syntax and one for the rest.
But it would leave this as it is for now, merge it and then work on those in a second PR

WDY

josevalim commented:
If we are worried about the table size we can also break the table per application. Although that may not help much.

josevalim commented on lib/elixir/pages/Deprecations.md:
```diff
@@ -0,0 +1,64 @@
+`\x{X*}` inside strings/sigils/charlists | [v1.3] | `\uXXXX`; `\u{X*}` | v1.1
+Map or dictionary as second argument in `Enum.group_by/3` | [v1.3] | Pass a function | v1.3
+Non-map as second argument in `URI.decode_query/2` | [v1.3] | Use a map | v1.0
+`Dict` behaviour          | [v1.2]        | *None*                  | *None*
```
The replacement is the `Map` data structure (similar to Set being the MapSet data structure).

eksperimental replied:
done

josevalim commented:
@eksperimental I gave another pass and found only one place where we could change `None` to something else. I have added it as a comment. :heart:

eksperimental commented:
I have merged the last two columns.
This is ready for a final review/merging.
thank you everyone for your input

whatyouhide commented:
Thanks! 💟
